allennlp/training/scheduler.py

```diff
@@ -0,0 +1,75 @@
+from typing import Dict, Any
+
+import torch
+
+
+class Scheduler:
+    """
+    A ``Scheduler`` is a generalization of PyTorch learning rate schedulers.
+
+    A scheduler can be used to update any field in an optimizer's parameter groups,
+    not just the learning rate.
+
+    During training using the AllenNLP `Trainer`, this is the API and calling
+    sequence for ``step`` and ``step_batch``::
+
+       scheduler = ... # creates scheduler, calls self.step(epoch=-1) in __init__
+
+       batch_num_total = 0
+       for epoch in range(num_epochs):
+           for batch in batchs_in_epoch:
+               # compute loss, update parameters with current learning rates
+               # call step_batch AFTER updating parameters
+               batch_num_total += 1
+               scheduler.step_batch(batch_num_total)
+           # call step() at the END of each epoch
+           scheduler.step(validation_metrics, epoch)
+    """
+
+    def __init__(self,
+                 optimizer: torch.optim.Optimizer,
+                 param_group_field: str,
+                 last_epoch: int = -1) -> None:
+        self.optimizer = optimizer
+        self.param_group_field = param_group_field
+        self._initial_param_group_field = f"initial_{param_group_field}"
+        if last_epoch == -1:
+            for i, group in enumerate(self.optimizer.param_groups):
+                if param_group_field not in group:
+                    raise KeyError(f"{param_group_field} missing from param_groups[{i}]")
+                group.setdefault(self._initial_param_group_field, group[param_group_field])
+        else:
+            for i, group in enumerate(self.optimizer.param_groups):
+                if self._initial_param_group_field not in group:
+                    raise KeyError(f"{self._initial_param_group_field} missing from param_groups[{i}]")
+        self.base_values = [group[self._initial_param_group_field] for group in self.optimizer.param_groups]
+        self.step(epoch=last_epoch)
```

Comment:
This might account for the difference in the trainer test -- in the pytorch base `torch.optim.lr_scheduler._LRScheduler` this line is `self.step(last_epoch + 1)`.

Reply:
@matt-peters yea, I think that the code / test was wrong before due to the way PyTorch LR schedulers are implemented. I left a note in the docstring explaining it. I think GitHub is having some issues RN though so my latest commits are not showing up

Reply:
The commits are showing up now, it was just slow. Sorry I didn't see your comment about holding off on the momentum schedulers yet. But I still haven't touched the trainer class